core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -66,140 +71,234 @@ class LogManager(val logDirs: Array[File],
   private val logs = new Pool[TopicPartition, Log]()
   private val logsToBeDeleted = new LinkedBlockingQueue[Log]()
 
-  createAndValidateLogDirs(logDirs)
-  private val dirLocks = lockLogDirs(logDirs)
-  private val recoveryPointCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
-  private val logStartOffsetCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+  private val _liveLogDirs: ConcurrentLinkedQueue[File] = createAndValidateLogDirs(logDirs, initialOfflineDirs)
+
+  def liveLogDirs: Seq[File] = {
+    if (_liveLogDirs.size() == logDirs.size)
+      logDirs
+    else
+      _liveLogDirs.asScala.toSeq
```

Comment:
I think it is probably OK to keep the current code. It is true that a log failure may happen right after the check in line 77. In this case the the caller may try to access that log directory that just became offline. However, there is no way to prevent this from happening since it is always possible for a log directory to become offline immediately before (or while) the caller tries to access it.

Thus caller code always needs to handle the possibility that log directory was removed from the state (e.g. `LogManager.recoveryPointCheckpoints`). Currently the only regular callers of `liveLogDirs()` are those checkpoint routines. The code has handled with e.g. `this.recoveryPointCheckpoints.get(dir).foreach(...)` to avoid NullPointerException.